File: converter.py
import os
import sys
import click
import pandas as pd
import subprocess
import pathlib
import csv
import pprint

use_pandas = False

## Will make this configurable
changestovcf_exec = './changestovcf.pl'

## Will make this configurable
## Can make this file type specific
##
COLUMNS = ['ChangeUID', 'IndexString', 'Coverage', 'MinQualityScore', 'MaxQualityScore', 'AverageQualityScore',
           'DistinctCoverage', 'MutPct', 'Chrom', 'Start', 'End', 'BaseFrom', 'BaseTo', 'LookupKey', 'ENSG',
           'GeneName']

QUALIFIED_COLUMNS_LOOKUP = None

## Will make this configurable
## The number of rows that are considered the header portion
## of the file and should be stripped prior to invoking the
## changestovcf executable.
HEADER_ROW_COUNT = 1

def get_qualified_columns_lookup():
    """
    """
    lookup = {}

    for column_name in COLUMNS:
        lookup[column_name] = True

    return lookup

def filter_file_using_pandas(infile, outfile, verbose):

    filtered_columns_df = pd.read_table(infile, sep='\t', low_memory=False, usecols=COLUMNS)

    if verbose:
        print(filtered_columns_df)

    filtered_columns_df.to_csv(outfile, index=False)

    print("Wrote output file CSV file '%s'" % outfile)


def filter_file_without_pandas(infile, outfile, verbose):

    line_ctr = 0;

    qualified_column_lookup = get_qualified_columns_lookup()

    qualified_column_number_lookup = {}

    qualified_column_number_list = []

    filtered_record_list = []

    qualified_header_list = []

    # pp = pprint.PrettyPrinter(width=41, compact=True)

    # pp.pprint(qualified_column_lookup)

    with open(infile, 'r') as csvfile:

        reader = csv.reader(csvfile, delimiter='\t')

        row_ctr = 0

        for row in reader:

            row_ctr += 1

            # print("Row number %d" %  row_ctr)

            if row_ctr == 1:

                # This is the header so want to derive the column numbers for our
                # fields of interest.
                header_ctr = 0;

                for header in row:

                    if header in qualified_column_lookup:

                        qualified_column_number_lookup[header_ctr] = header

                        qualified_column_number_list.append(header_ctr)

                        qualified_header_list.append(header)

                    header_ctr += 1

                # pp.pprint(qualified_column_number_list)
                # pp.pprint(qualified_column_number_lookup)
                continue

            else:

                # print("Processing line %d" % row_ctr)

                record = []

                for field_num in qualified_column_number_list:

                    record.append(row[field_num])

                filtered_record_list.append(record)


    with open(outfile, 'w') as csvfile:

        writer = csv.writer(csvfile, delimiter='\t')

        writer.writerow(qualified_header_list)

        for row in filtered_record_list:

            writer.writerow(row)

    print("Wrote output file CSV file '%s'" % outfile)




def get_file_line_count(fname):
    """

    :param fname: file for which line count should be determined

    :return: int
    """
    with open(fname) as f:
        for i, l in enumerate(f):
            pass
    return i + 1


def run_changestovcf(infile, outfile, count, verbose):

    cmd = "tail -" + str(count) + ' ' + infile + ' | perl ' + changestovcf_exec + ' > ' + outfile
    if verbose:
        print("Will attempt to execute : '%s'" % cmd)

    p = subprocess.Popen(cmd, shell=True)

    (stdout, stderr) = p.communicate()

    p_status = p.wait()

    if p_status == 0:
        print("VCF file '%s' is ready" % outfile)
    else:
        print("Encountered some problem during invocation of '%s' : %s " % (changestovcf_exec, stderr))

def create_infile_symlink(infile, symlink_file):
    """
    :param infile: The source file
    :param symlink_file: The destination file
    """

    try:
        os.symlink(infile, symlink_file)
    except:
        print("Encountered some error while attempt to create symlink %s -> %s" % (infile, symlink_file))
        raise



@click.command()
@click.argument('infile')
@click.argument('sample_id')
@click.option('--verbose', default=False, is_flag=True)
@click.option('--create_symlink', default=True, is_flag=True)
@click.option('--outdir', default='./')
@click.option('--header_row_count', default=1)
def main(infile, sample_id, verbose, create_symlink, outdir, header_row_count):
    """

    INFILE : The input file to be processed.

    SAMPLE_ID: The sample identifier.


    :return: None


    """
    assert isinstance(infile, str)
    assert isinstance(sample_id, str)
    assert isinstance(header_row_count, int)

    if not outdir == './':
        pathlib.Path(outdir).mkdir(parents=True, exist_ok=True)
        if verbose:
            print("output directory '%s' was created" %  outdir)


    if not os.path.isfile(infile):
        print("'%s' is not a file" % infile)
        sys.exit(1)


    if verbose:
        print("The input file is %s" % infile)
        print("The sample_id is %s" % sample_id)

    infile_basename = os.path.basename(infile)

    infile_basename_lc = infile_basename.lower()

    file_type = 'Unknown'

    if "allchanges.txt" in infile_basename_lc:
        file_type = 'Unfiltered'
    elif "plasmachanges.txt" in infile_basename_lc:
        file_type = 'Filtered'
    elif "vv_changes.txt" in infile_basename_lc:
        file_type = 'MAF-Filtered'
    else:
        print("Unsupported file type")

    if verbose:
        print("Found file type '%s'" % file_type)


    if verbose:
        print("Including only the following column names:")
        for name in COLUMNS:
            print(name)

    # Retrieve subset of columns filtered by column
    # names specified in the columns list.

    outfile = os.path.join(outdir, sample_id + "_" + file_type + ".csv")

    if use_pandas:
        filter_file_using_pandas(infile, outfile, verbose)
    else:
        filter_file_without_pandas(infile, outfile, verbose)


    line_count = get_file_line_count(infile)

    tail_count = line_count -1

    out_vcf_file = os.path.join(outdir, sample_id + "_" + file_type + ".vcf")

    run_changestovcf(outfile, out_vcf_file, tail_count, verbose)

    if create_symlink:

        symlink_file  = outdir + '/' + infile_basename

        create_infile_symlink(infile, symlink_file)


if __name__ == "__main__":
    main()

File: converter.pl
#!/usr/bin/env perl
use strict;
use Carp;
use Term::ANSIColor;
use Pod::Usage;
use File::Path;
use File::Basename;
use File::Copy;
use Getopt::Long qw(:config no_ignore_case no_auto_abbrev);


use constant TRUE => 1;

use constant FALSE => 0;

use constant DEFAULT_CREATE_SYMLINK => FALSE;

use constant DEFAULT_HEADER_ROW_COUNT => 1;

use constant DEFAULT_VERBOSE => TRUE;

use constant DEFAULT_OUTDIR => '/tmp/' . File::Basename::basename($0) . '/' . time();


## Will make this configurable
my $changestovcf_exec = './changestovcf.pl';

## Will make this configurable
## Can make this file type specific
##
my $COLUMNS = ['ChangeUID', 'IndexString', 'Coverage', 'MinQualityScore', 'MaxQualityScore', 'AverageQualityScore',
	       'DistinctCoverage', 'MutPct', 'Chrom', 'Start', 'End', 'BaseFrom', 'BaseTo', 'LookupKey', 'ENSG',
	       'GeneName'];

my $QUALIFIED_COLUMNS_LOOKUP = {};

## Will make this configurable
## The number of rows that are considered the header portion
## of the file and should be stripped prior to invoking the
## changestovcf executable.
my $HEADER_ROW_COUNT = 1;

my (
    $help,
    $man,
    $infile,
    $sample_id,
    $verbose,
    $create_symlink,
    $outdir,
    $header_row_count
    );

my $results = GetOptions (
    'infile=s'           => \$infile,
    'sample_id=s'        => \$sample_id,
    'verbose=s'          => \$verbose,
    'create_symlink=s'   => \$create_symlink,
    'help|h'             => \$help,
    'man|m'              => \$man,
    'outdir=s'           => \$outdir,
    'header_row_count=s' => \$header_row_count,
    );

&checkCommandLineArguments();

&main();


##------------------------------------------------------
##
##  END OF MAIN -- SUBROUTINES FOLLOW
##
##------------------------------------------------------

sub checkCommandLineArguments {

    if ($man){
    	&pod2usage({-exitval => 1, -verbose => 2, -output => \*STDOUT});
    }

    if ($help){
    	&pod2usage({-exitval => 1, -verbose => 1, -output => \*STDOUT});
    }

    my $fatalCtr=0;

    if (!defined($infile)){

        printBoldRed("--infile was not specified");

    	$fatalCtr++;
    }

    if (!defined($sample_id)){

        printBoldRed("--sample_id was not specified");

    	$fatalCtr++;
    }


    if (!defined($verbose)){

        $verbose = DEFAULT_VERBOSE;

        printYellow("--verbose was not specified and therefore was set to default '$verbose'");
    }

    if (!defined($header_row_count)){

        $header_row_count = DEFAULT_HEADER_ROW_COUNT;

        printYellow("--header_row_count was not specified and therefore was set to default '$header_row_count'");
    }

    if (!defined($create_symlink)){

        $create_symlink = DEFAULT_CREATE_SYMLINK;

        printYellow("--create_symlink was not specified and therefore was set to default '$create_symlink'");
    }

    if (!defined($outdir)){

    	$outdir = DEFAULT_OUTDIR;

        printYellow("--outdir was not specified and therefore was set to default '$outdir'");
    }


    $outdir = File::Spec->rel2abs($outdir);

    if (!-e $outdir){

        mkpath ($outdir) || die "Could not create output directory '$outdir' : $!";

        printYellow("Created output directory '$outdir'");

    }

    if ($fatalCtr> 0 ){

    	die "Required command-line arguments were not specified\n";
    }
}

sub printBoldRed {

    my ($msg) = @_;
    print color 'bold red';
    print $msg . "\n";
    print color 'reset';
}

sub printYellow {

    my ($msg) = @_;
    print color 'yellow';
    print $msg . "\n";
    print color 'reset';
}

sub printGreen {

    my ($msg) = @_;
    print color 'green';
    print $msg . "\n";
    print color 'reset';
}

sub main {

    if (!-e $infile){
    	die "infile '$infile' does not exist";
    }

    if ($verbose){
        print("The input file is '$infile'\n");
    	print("The sample_id is '$sample_id'\n");
    }

    my $infile_basename = File::Basename::basename($infile);

    my $file_type = 'Unknown';

    if ($infile_basename =~ m/allchanges\.txt$/i){

        $file_type = 'Unfiltered';
    }
    elsif ($infile_basename =~ m/plasmachanges\.txt$/i){

        $file_type = 'Filtered';
    }
    elsif ($infile_basename =~ m/vv_changes\.txt/i){

        $file_type = 'MAF-Filtered';
    }
    else {

        print("Unsupported file type '$infile'\n");
    	exit(0);
    }

    if ($verbose){
        print("Found file type '$file_type'\n");
    }

    if ($infile_basename =~ m/\.gz$/){
        $infile = &uncompress_file($infile);
    }

    if ($verbose){
        print("Including only the following column names : " . join(', ', @{$COLUMNS})) . "\n";
    }


    # Retrieve subset of columns filtered by column
    # names specified in the columns list.

    my $outfile = $outdir . '/' . $sample_id . '_' . $file_type . '.csv';

    &filter_file_without_pandas($infile, $outfile);

    my $line_count = &get_file_line_count($infile);

    my $tail_count = $line_count - 1;

    my $out_vcf_file = $outdir . '/' . $sample_id . "_" . $file_type . '.vcf';

    &run_changestovcf($outfile, $out_vcf_file, $tail_count);

    if ($create_symlink){

        my $symlink_file  = $outdir . '/' . $infile_basename;

        &create_infile_symlink($infile, $symlink_file);
    }
}

sub execute_cmd {

    my ($cmd) = @_;

    if ($verbose){
        print("Will attempt to execute '$cmd'\n");
    }

    eval {
    	qx($cmd);
    };

    if ($?){
    	confess "Encountered some error while attempting to execute '$cmd' : $@ $!";
    }

}

sub uncompress_file {

    my ($infile, $verbose) = @_;

    my $cmd = "gunzip -f " . $infile;

    if ($infile =~ m/\.gz$/){
    	$infile =~ s/\.gz$//;
    }

    &execute_cmd($cmd);

    return $infile;
}

sub get_qualified_columns_lookup {

    my $lookup = {};

    for my $column_name (@{$COLUMNS}){
        $lookup->{$column_name}++;
    }

    return $lookup;
}

sub filter_file_without_pandas {

    my ($infile, $outfile) = @_;

    my $line_ctr = 0;

    my $qualified_column_lookup = &get_qualified_columns_lookup();

    my $qualified_column_number_lookup = {};

    my $qualified_column_number_list = [];

    my $filtered_record_list = [];

    my $qualified_header_list = [];


    open (INFILE, "<$infile") || die "Could not open input file '$infile' in read mode : $!";


    my $already_found_lookup = {};

    my $row_ctr = 0;

    while ( my $line = <INFILE>){

    	chomp $line;

    	my @parts = split("\t", $line);

    	my $row = \@parts;

    	$row_ctr++;

    	if ($row_ctr == 1){

    	    # This is the header so want to derive the column numbers for our
    	    # fields of interest.

    	    my $header_ctr = 0;

    	    for my $header (@{$row}){

        		if (exists $qualified_column_lookup->{$header}){

        		    if (! exists $already_found_lookup->{$header}){

            			$qualified_column_number_lookup->{$header_ctr} = $header;

            			push(@{$qualified_column_number_list}, $header_ctr);

            			push(@{$qualified_header_list}, $header);

            			$already_found_lookup->{$header}++;
        		    }
        		}

        		$header_ctr++;

    	    }
    	}
    	else {

    	    my $record = [];

    	    for my $field_num (@{$qualified_column_number_list}){

        		push(@{$record}, $row->[$field_num]);
    	    }

    	    push(@{$filtered_record_list}, $record);
    	}
    }



    open (OUTFILE, ">$outfile") || die "Could not open output file '$outfile' in write mode : $!";

    print OUTFILE join(',', @{$qualified_header_list}) . "\n";

    for my $row (@{$filtered_record_list}){

    	print OUTFILE join(',', @{$row}) . "\n";
    }

    print("Wrote output file CSV file '$outfile'\n");
}

sub get_file_line_count {

    my ($fname) = @_;

    open (INFILE, "<$infile") || die "Could not open input file '$infile' in read mode : $!";

    my $row_ctr = 0;

    while ( my $line = <INFILE>){

    	chomp $line;
    	$row_ctr++;
    }

    return $row_ctr;
}


sub run_changestovcf {

    my ($infile, $outfile, $count) = @_;

    my $cmd = "tail -" . $count . ' ' . $infile . ' | perl ' . $changestovcf_exec . ' > ' . $outfile;

    &execute_cmd($cmd);

    print("VCF file '$outfile' is ready\n");
}


sub create_infile_symlink {

    my ($infile, $symlink_file) = @_;

    my $cmd = "ln -s $infile $symlink_file";

    &execute_cmd($cmd);
}


